zerver/views/auth.py

```diff
@@ -290,6 +290,14 @@ def login_or_register_remote_user(request: HttpRequest, result: ExternalAuthResu
       are doing authentication using the mobile_flow_otp or desktop_flow_otp flow.
     """
     user_profile = result.user_profile
+
+    action_key = result.data_dict.get('action_key')
+    if action_key:
+        if user_profile is None:
```

Comment:
if you put this block a few lines below, it would get much simpler, as whether user needs to register is checked and handled right below (and it handles the mirror dummy case too, which also requires registering and was missed here.

Reply:
Do we want to raise a JSON error or just take the user to registration page? Raising an error seems right to me as it could avoid some confusion if the user accidentally logs in with some other account. So, for now I made it raise json error. we can change it if you think it's not ideal.

Reply:
Hmm okay I think you're right, the user might get confused and log in with the new email rather than old, taking them to registration would be weird and an error is better. Json error is not ideal, it's be better experience to take the user to the login page again with the action key and an appropriate message - but doing that is more work so let's keep just the json error for now and make error handling nicer in followups. 

But `user_profile.is_mirror` needs to be accounted for, and probably you can make this login join with that block below that has the same `if` in a way that'll read nicer.

Reply:
Yes, I added an `if` inside `if user_profile is None or user_profile.is_mirror_dummy:` to check for action key and raise error.  Now looking at it again, I felt like keeping `action_key` related code as one block in the code would look clean. So thought I'd modify like this, but my doesn't let me(see [this post](https://chat.zulip.org/#narrow/stream/92-learning/topic/mypy.20error/near/968696) in chat.zulip.org)
```python
....
    user_needs_to_register = user_profile is None or user_profile.is_mirror_dummy

    action_key = result.data_dict.get('action_key')
    if action_key:
        if user_needs_to_register:
            raise JsonableError(_("Incorrect login. Login with your old email to change your email address"))
        return handle_action_key(request, user_profile, action_key)

    if user_needs_to_register:
        return register_remote_user(request, result)
    # Otherwise, the user has successfully authenticated to an
...
```